subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultVisualStudioLocator.java

```diff
@@ -105,105 +100,149 @@ public SearchResult locateDefaultVisualStudioInstall(File candidate) {
-                LOGGER.debug("Found Visual C++ {} at {}", valueName, visualCppDir);
-                VersionNumber version = VersionNumber.parse(valueName);
-                VisualCppInstall visualCpp = buildVisualCppInstall("Visual C++ " + valueName, visualStudioDir, visualCppDir, version);
-                VisualStudioInstall visualStudio = new VisualStudioInstall(visualStudioDir, visualCpp);
-                foundInstalls.put(visualStudioDir, visualStudio);
-            } else {
-                LOGGER.debug("Ignoring candidate Visual C++ directory {} as it does not look like a Visual C++ installation.", visualCppDir);
-            }
+        if (isValidInstall(install)) {
+            LOGGER.debug("Found Visual C++ {} at {}", install.getVersion(), visualCppDir);
+            VersionNumber version = install.getVersion();
+            VisualCppInstall visualCpp = buildVisualCppInstall("Visual C++ " + install.getVisualCppVersion(), visualStudioDir, visualCppDir, install.getVisualCppVersion(), install.getCompatibility());
+            VisualStudioInstall visualStudio = new VisualStudioInstall("Visual Studio " + version, visualStudioDir, version, visualCpp);
+            foundInstalls.put(visualStudioDir, visualStudio);
+        } else {
+            LOGGER.debug("Ignoring candidate Visual C++ directory {} as it does not look like a Visual C++ installation.", visualCppDir);
         }
     }
 
     private void locateInstallInPath() {
-        File compilerInPath = os.findInPath(COMPILER_FILENAME);
+        File compilerInPath = os.findInPath(LEGACY_COMPILER_FILENAME);
         if (compilerInPath == null) {
             LOGGER.debug("No visual c++ compiler found in system path.");
             return;
         }
 
-        File visualCppDir = FileUtils.canonicalize(compilerInPath.getParentFile().getParentFile());
-        if (!isVisualCpp(visualCppDir)) {
-            visualCppDir = visualCppDir.getParentFile();
-            if (!isVisualCpp(visualCppDir)) {
+        VisualStudioMetadata install = versionDeterminer.getVisualStudioMetadataFromCompiler(compilerInPath);
+
+        if (install != null) {
+            File visualCppDir = install.getVisualCppDir();
+            if (!isValidInstall(install)) {
                 LOGGER.debug("Ignoring candidate Visual C++ install for {} as it does not look like a Visual C++ installation.", compilerInPath);
                 return;
             }
+            LOGGER.debug("Found Visual C++ install {} using system path", visualCppDir);
+
+            File visualStudioDir = install.getInstallDir();
+            if (!foundInstalls.containsKey(visualStudioDir)) {
+                String displayVersion = install.getVersion() == VersionNumber.UNKNOWN ? "from system path" : install.getVersion().toString();
+                String displayCppVersion = install.getVisualCppVersion() == VersionNumber.UNKNOWN ? "from system path" : install.getVisualCppVersion().toString();
```

Comment:
we should reject a VS install for which we don't know the version. This will be important later when we attempt to build for specific platforms or consume prebuilt binaries.

Reply:
So the only time we would do this is if we find a version on the path or the user specifies an installation that can't be found from `vswhere` or the registry.  Right now, `vswhere` and the registry are our only reliable mechanisms for discovering version information.  Since we only look at the path if we can't find an installation in the other ways, it's always an unknown version.  So, saying we should reject unknown versions is essentially saying we should not look for an installation on the path (at least currently).

What we could do is look for other ways to infer the VisualCpp version from an installation directory.  With VS2017, this is baked into the directory structure, so we should never have an unknown version for it.  For previous versions, the compiler [seems to return a wildly different version](https://stackoverflow.com/questions/1233312/finding-version-of-microsoft-c-compiler-from-command-line-for-makefiles), but the linker seems to display a version that is consistent with the version from the registry, so probing `link.exe` seems promising (I only looked at VS2015 and VS2013, though).  I'll create an issue to investigate this further and see if we can find a reliable strategy.

Until we have such a strategy, do you feel like we should just ignore these versions (and therefore ignore installs <VS2017 on the path)?

Reply:
When would it _actually_ happen that we wouldn't have already discovered the install provided on the path? I don't think we need to probe any installs and we can reject anything for which we don't know the version as we'll almost certainly have discovered it already.

I'd make it strict until we hear otherwise.

Reply:
Sounds good.  Yeah, no idea if this is a real use case or not - it's just the way it worked before.

I've changed it so that we reject any install (on the path or user-specified) where we can't determine the version of VisualCpp.